### ISSUE TITLE
[Discussion] Session Stickyness

### ISSUE THREAD

**Issue by mzinnoecker** [abap2UI5/abap2UI5#1327]
Hello!

@oblomov-dev this is a great project. Just wanted to mention that. Thanks!

Have you ever thought of wether we would like to have Sessions/some specific apps stateful (i mean really stateful)? (Also called session stickyness nowadays in context of RAP)

I am thinking of porting some heavy duty GUI transactions and as this is mainly legacy code… you know, very hard for having all the stuff serializable and/or at the end stateless.

some articles about that:
https://sap.github.io/odata-vocabularies/vocabularies/Session.html

https://help.sap.com/doc/saphelp_autoid2007/2007/en-US/05/f9e8005d944049b15eac7008238c3f/content.htm?no_cache=true

I would think it would need some cookie handling on the client side, session termination/open behavior and handling on the SICF handler.
Maybe it makes sense to have that feature in the framework… your thougts?
Just wanted to discuss before elaborating further…
Best,
Michael

**Comment by oblomov-dev** [comment]
HI @mzinnoecker,

thanks for writing this, very interesting, never looked further in this direction till now. Work mostly on S/4 Public at the moment so the fake "stateful" way with serialization was always enough for me. 

But i worked a lot on old onPrem systems and with stateful sessions in the past and yes the abap stack is full of stateful legacy code which could be nice being able to call via stateful abap2UI5. It was also great fun making abap2UI5 compatible for lower releases, so more use cases and more compatibility is always good for the project i think :)

Unfortunately I never developed a stateful handler by myself. Are you familiar with this? Maybe adding some logic in JS before the HTTP Post and some ABAP Logic in the handler is needed? Feel free to give it a try and i will help you as much as i can when you need information about the code of abap2UI5.

Would be nice adding this feature to abap2UI5!

Best regards,
oblomov

**Comment by mzinnoecker** [comment]
Hey,
thanks for the answer. 
Did work with client side implementation of Soft state OData some time ago in a freestyle ui5 app which is quite comparable from a clients perspective.
Also have some understanding of the rest handler implementation and we can have a look at odata v4 implementation in newer releases too, as they(SAP) does support session stickyness(stateful) too.
Am a bit limited in the next 2 weeks, then probably will do some prototyping, except anyone will do that before me. 
will come back :)
Best,
Michael

**Comment by oblomov-dev** [comment]
nice, this is a good plan 👍

**Comment by mzinnoecker** [comment]
hey @oblomov-dev 
would you mind sharing the UI5 (vscode?) project for the ABAP2UI5 skeleton? (or point me to where this is located)
I guess you did not develop that in Z2UI5_CL_CORE_HTTP_GET purely by string concatenation?
I would first like to do that cookie/session handling in a proper IDE and then move that to the ABAP class
Am i guessing right?
Thanks!
Best,
Michael

**Comment by mzinnoecker** [comment]
hey @oblomov-dev 
would you mind sharing the UI5 (vscode?) project for the ABAP2UI5 skeleton? (or point me to where this is located)
I guess you did not develop that in Z2UI5_CL_CORE_HTTP_GET purely by string concatenation?
I would first like to do that session handling termination in a proper IDE and then move that to the ABAP class
Am i guessing right that there may be some local project?
Thanks!
Best,
Michael

**Comment by oblomov-dev** [comment]
hehe yes somehow most of the code is really created purely inside the string concatenation, so you will see that the frontend JS code is unfortunately not a beauty.

but in the beginning i used this repository when i had to make bigger changes:
https://github.com/abap2UI5/dev-js_files_of_abap2UI5/

but its a outdated now and there were some manual steps to do. i will check this again if i can make it run and give you an update.

**Comment by mzinnoecker** [comment]
Maybe you can help me that way too?
I need to do a request for terminating backend session when an App is closed, navigated away, browser window closed,....
In a "normal" UI5 application i did that in Component.js destroy( ) function.
I may be wrong, but you do not work with Components right?
Just wanting to save some hours for digging... do you see some hook in the current implementation where i can react on "navigate away from app" (browser window close can be handled with controller -> destroy( ) ) 
Thanks for the help

**Comment by mzinnoecker** [comment]
Hey @oblomov-dev  maybe you can help me out a bit...
Forget the local project, i saw that it's sufficient for me to have a look at the network tab in chrome and do the work also via string concat.
I need to do some request when the user navigates away from the app/closes the browser windows/... so i am currently searching for an alternative way of doing it via Component.destroy( ).
I saw that you are not using Components
Controller->onExit does not behave the same. Do you see another way except from doing it in controller.onexit() and also window.beforeunload?
Thanks,
Michael

**Comment by oblomov-dev** [comment]
yes so far i did not find a way of how to use a component container in a one page html application... 🤷‍♂️

From UI5 events, i also don't know anything better, so i guess using the window.beforeunload is the best option. we used the same to check for "dirty" data as you can see here:
https://github.com/abap2UI5/abap2UI5/blob/main/src/01/04/z2ui5_cl_cc_dirty.clas.abap#L37

**Comment by mzinnoecker** [comment]
Thanks for the answer! Will think about it.
wondered why you did it not the same as the index.html fiori tools generate? (For standalone usage w/o launchpad)

here is an example, my component is „qpdsexample“

`<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>App Title</title>
    <style>
        html, body, body > div, #container, #container-uiarea {
            height: 100%;
        }
    </style>
    <script
        id="sap-ui-bootstrap"
        src="resources/sap-ui-core.js"
        data-sap-ui-theme="sap_horizon"
        data-sap-ui-resourceroots='{
            "qpdsexample": "./"
        }'
        data-sap-ui-oninit="module:sap/ui/core/ComponentSupport"
        data-sap-ui-compatVersion="edge"
        data-sap-ui-async="true"
        data-sap-ui-frameOptions="trusted"
    ></script>
</head>
<body class="sapUiBody sapUiSizeCompact" id="content">
    <div
        data-sap-ui-component
        data-name="qpdsexample"
        data-id="container"
        data-settings='{"id" : "qpdsexample"}'
        data-handle-validation="true"
    ></div>
</body>
</html>`

**Comment by oblomov-dev** [comment]
mhh... but how to tell the ui5 bootstrapping not to search for the compoent file somewhere? your file for example creates this error:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/49ad2fba-d62a-4f66-bc6c-ed2ae7d63ccc">
maybe there is an easy solution, would not tell myself an ui5 expert, but i don't know how to tell the UI5 framework that we define the component container directly on the index.html.
we had the same struggle with the manifest, for example here:
https://github.com/abap2UI5/abap2UI5/issues/1147
hints always welcome :)

**Comment by mzinnoecker** [comment]
Ok got it… may have an idea. Will come back to you next week

**Comment by mzinnoecker** [comment]
Hello @oblomov-dev 
did open pull request #1433, maybe you want to have a look at this?
It does implement session stateful handling and a suggestion for adding the Component. Only as a first version, please review... And if OK maybe we want to transition to fully standard? (use of Component.destroy() in the implementation for Session Stateful handling does not work until transitioned)
Thanks

**Comment by oblomov-dev** [comment]
first of all, thank your for the PR and putting effort and your knowledge into the project! I am sure it took a little bit to get into the code and you needed to make several changes, so great job :)

demo works on my side and all test with the other samples, launchpad etc too.  
when i saw it right you also didnt need to change the frontend request, just a change of the backend server is enough. never expected to get it as easy like this. i am sure it will be a nice feature for a lot of people, there is too much legacy code out there which can be called stateful now. 

to the other aspects:
i have in mind to bring abap2UI5 always as close (as this approach allows) to the standard UI5 behaviour. i try to reduce any unallowed API calls etc. but as i said in the comment before, my home is ABAP so my UI5 knowledge is limited. its very nice to see that you find a way to include component container and manifest. changes in this direction is exactly what the project needs! 

so should we extend the manifest as a next step that Component.destroy will be triggered correctly or what do you recommend?

you can just take the test-comp-stateful branch and push the changes you need. or you let me know what i can do on my side and then i give it a try.

thanks again & best regards

**Comment by mzinnoecker** [comment]
Hey,
thanks for your adaptions too and also for the kind words... Glad to see that everyone can bring his/her knowledge to get the project forward :)

Regarding:
"so should we extend the manifest as a next step that Component.destroy will be triggered correctly or what do you recommend?"
YES, i would like to add routing configuration and getting rid of the explicit placeAt() -> so component does take care of the View etc.
This would allow adding the flexibility things and maybe more sophisticated dynamic route configurations per app in the future 😄 
But thats a future thing, first lets focus on full functional stateful handling :)
I am a bit limited but i will try to find some free slots this/next week
Best,
Michael

**Comment by oblomov-dev** [comment]
somehow this project got bigger than expected, so every helping hand & knowledge is needed :)

yes let’s finish the stickyness first and then step by step. all changes sound good, you can also create separated issues if needed. 

no problem, take your time with the adjustments, I will merge the actual changes to main. although they are not completely finished they do not break anything and we have the chance to get early feedback if someone runs into problems and we overlooked something.

**Comment by choper725** [comment]
sounds intresting !! 🙌

**Comment by oblomov-dev** [comment]
@choper725 adjust the handler as described here, and then check sample 137. everything merged now.

**Comment by choper725** [comment]
thx guys.. excellent ! let me know if i can help in anything..

**Comment by christianguenter2** [comment]
Amazing new feature 👍  @mzinnoecker great to have you on board!

**Comment by christianguenter2** [comment]
Next thing I'd like is to support durable enqueue locks with sticky sessions. Currently it's not possible because of these `ROLLBACK WORK`s here.
https://github.com/abap2UI5/abap2UI5/blob/1fba55457c6af0a7f38e425dad9d0bc72ebb1453/src/01/02/z2ui5_cl_core_http_post.clas.abap#L137-L139

Maybe we can omit them if stateful session is active?

**Comment by mzinnoecker** [comment]
I would love to see that! 😃

**Comment by oblomov-dev** [comment]
yes no problem, let's continue here with the changes above and i will create a separated issue to take a look to the rollback thing.

**Comment by mzinnoecker** [comment]
Hey @oblomov-dev
did open a pull request #1439
This looks like as a small refactoring, but can potentially have huge (negative) impact. Though i do not expect any, i did already test some things... but would be happy if you can especially have a look at the more complex things like nested views etc :)

What i did:
- added Component support and refactored View Placement into Root View of Component (now the abap2ui5 takes part on the standard lifecycle)
- changed the stateful behavior from cookie to request/response headers... The cookies are shared across different tabs in a browser instance and thus across different apps... that resulted in weird behavior. Now its just like SAP does with the OData handler (header based instead of cookie based)
- unfortunately the Component.exit is only called within Launchpad :( (i did not know that) and thus i needed to add an listener in Component.js for "pagehide" -> but nevertheless adding the Component i think has huge advantages for the future (we could potentially make use of routers somehow)

It would be great if you can have a look at :)
Thanks!

**Comment by mzinnoecker** [comment]
Did forget to mention that the REST handler will have a bit more lines to support all these...
any suggestions appreciated (i got the point that you would not like to have the REST handler in GIT in order to also support cloud)

please adapt for your tests with the sample app
`
    DATA(lv_method) = server->request->get_method( ).
    IF lv_method = 'HEAD'.
      "terminate session, currently thats the only supported HEAD request
      server->set_session_stateful( stateful = if_http_server=>co_disabled ).
      RETURN.
    ENDIF.

    z2ui5_cl_http_handler=>main(
      EXPORTING
        body       = server->request->get_cdata( )
        config     = VALUE #( t_config = VALUE #( ( n = `src` v = `/sap/public/bc/ui5_ui5/resources/sap-ui-core.js` ) ) )
      IMPORTING
        attributes = DATA(attributes)
      RECEIVING
        result     = DATA(response_body) ).
    server->response->set_cdata( response_body ).
    server->response->set_header_field( name = `cache-control` value = `no-cache` ).
    server->response->set_status( code = 200 reason = `success` ).

    "transform cookie to header based contextid handling
    IF attributes-stateful-switched = abap_true.
      server->set_session_stateful( stateful = attributes-stateful-active ).
      IF server->request->get_header_field( 'sap-contextid-accept' ) = 'header'.
        server->response->get_cookie(
          EXPORTING
            name  = 'sap-contextid'
          IMPORTING
            value = DATA(lv_contextid) ).
        IF lv_contextid IS NOT INITIAL.
          server->response->delete_cookie( 'sap-contextid' ).
          server->response->set_header_field( name = 'sap-contextid' value = lv_contextid ).
        ENDIF.
      ENDIF.
    ELSE.
      lv_contextid = server->request->get_header_field( 'sap-contextid' ).
      IF lv_contextid IS NOT INITIAL.
        server->response->set_header_field( name = 'sap-contextid' value = lv_contextid ).
      ENDIF.
    ENDIF.`

**Comment by oblomov-dev** [comment]
wow thanks again! great work, checked with the samples and everything works fine 👌 

the only problems are the connectors, need to adjust the launchpad apps to make the new approach also run there. will therefore keep the branch for a while until i find time to fix that.

the component container is also now a good chance for some refactoring, will check my code if i can put a few things into the component container (device model etc).

good idea with changing to request/response headers, and yes a pity that Component.exit is only called in launchpad, but then we know now that a UI5 freestyle apps would also go with pagehide and are not better than abap2UI5 :)

the handler, yes so far i did not include it to the project, want to keep a single code line for both language flavours and i also think everyone customize it anyway with the possibilities described here:
https://github.com/abap2UI5/abap2UI5-documentation/blob/main/docs/handler_config.md

i will extend the documentation with a stateful session and maybe there is a good place for your new snippet.

**Comment by christianguenter2** [comment]
I get this dump when I access the first sticky examples via sample overview, navigate back and then switch to the other sticky sample.

![image](https://github.com/user-attachments/assets/1a01dd1c-38bd-4431-8268-8e02bd5cdfea)
![image](https://github.com/user-attachments/assets/6c7b035e-8c57-4300-bdf2-d2fb96b5ff50)
![image](https://github.com/user-attachments/assets/f5cb2651-5a4d-4d3d-beaa-15829f3fa339)
![image](https://github.com/user-attachments/assets/e79f046c-aca9-4d60-aad4-3e498fce6789)
![image](https://github.com/user-attachments/assets/7e5507cb-a099-44a4-8be0-a727dc9eebf9)

**Comment by oblomov-dev** [comment]
ah reproduced on my system. thanks for hunting the bug, its on my list now.

**Comment by oblomov-dev** [comment]
fixed now the bug with https://github.com/abap2UI5/abap2UI5/tree/test-components-views

additionally deactivated stateful sessions when leaving the app here:
https://github.com/abap2UI5/abap2UI5-samples/blob/main/src/z2ui5_cl_demo_app_137.clas.abap#L76

the navigation in abap2UI5 is done via a kind of chain with serialized apps on the database, avoiding commits here would be a very big change. but maybe having one LUW on app level is enough anyway?
i think with dynpros & screens there was also a commit done with every "call screen", so we would keep the traditional developer experience ;)

furthermore added an exception when starting more than one session (without ending before).
<img width="400" alt="image" src="https://github.com/user-attachments/assets/a937d72b-ff8e-4392-b3ad-848438de9036">

otherwise there is a dump, alternative would be just ignoring, but two session makes no sense to me?

give it a try, and let me know what you think.

**Comment by mzinnoecker** [comment]
Sounds good to me!
Thanks @christianguenter2 @oblomov-dev

**Comment by christianguenter2** [comment]
Did some tests, works exactly as expected. Thank you very much!

**Comment by christianguenter2** [comment]
When I close a sticky app (non FLP) irregularly e.g. by closing the browser, the session (and acquired locks) remains in the backend. Should be possible with `window.onbeforeunload`?

**Comment by mzinnoecker** [comment]
Hey @christianguenter2 
tbh, i hoped to not have some of this irregular circumstances. But its 100% sure to happen if you hope to not, right? 😄 

I checked the docs and the reason why i decided to go for `pagehide` was that `beforeunload` is not reliable according to the docs [here](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event). Now we learned that `pagehide` is not reliable too 😄

In my opinion we should make use of [visibilitychange event](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event), what do you think? (tbh, I am not that an expert in cross browser optimization)

**Comment by christianguenter2** [comment]
@mzinnoecker I'm not an expert either ;-)
I'm a big fan of trying things out and see whether they work or not. So let's give it try. 
Maybe also a combination of `pagehide` **and** `beforeunload` is possible?

**Comment by oblomov-dev** [comment]
i also have no better i ideas, its always a bit a mess with these browsers 😅 give it a try sounds good and if issues come up in the future we can just rethink it one more time (continuous improvement).
@christianguenter2 maybe the dirty data handling should use the same events like the session?

**Comment by christianguenter2** [comment]
> @christianguenter2 maybe the dirty data handling should use the same events like the session?

agree

**Comment by mzinnoecker** [comment]
PR #1440 
in my opinion the best we can have is a combination of `pagehide` for ios devices and `beforeunload` for the rest.
`visibilitychange` is not the event we are searching for, as it also gets triggered when browser window was minimized.

I did have some tests (browser close, tab close, navigate away, refresh) with chrome/firefox/safari on mac and windows... ios not tested yet but i understand that this maybe is not our main focus (thus should work)

**Comment by oblomov-dev** [comment]
merged. yes chrome/firefox/safari is more than enough. i guess sap only recommends chrome/chromium. thanks!

**Comment by christianguenter2** [comment]
Thx @mzinnoecker and @oblomov-dev works for me now 🚀

**Comment by oblomov-dev** [comment]
yes need to fix the launchpad connector first for the new component structure, but good to hear that the functionality itself works now 💪

**Comment by mzinnoecker** [comment]
Anyone has access to a s/4 public Cloud system and can check, if we can do something similar in the rest handler as we do on prem?
Just curious 🤔

**Comment by oblomov-dev** [comment]
check it, on an abap cloud the handler looks like this:
<img width="888" alt="Eclipse 2024-09-22 10 00 43" src="https://github.com/user-attachments/assets/7944997b-c56d-4f14-8476-5740cd70d80d">

<img width="520" alt="Eclipse 2024-09-22 10 01 03" src="https://github.com/user-attachments/assets/1467e9d5-fc4f-4aa3-8740-7509e8b76e1b">

<img width="493" alt="Eclipse 2024-09-22 10 00 55" src="https://github.com/user-attachments/assets/cc0d5988-9d3f-4604-85a2-923e1dbf578b">

but unfortunately I can't see any set_session_stateful 🤷‍♂️

one question:

if found time to look into the details of your change. very nice trick with sap.ui.require.preload! Did not know this, but i changed now everything to "simulate" the structure of a normal UI5 app:
https://github.com/abap2UI5/abap2UI5/blob/test-components-views/src/01/02/z2ui5_cl_core_http_get.clas.abap#L145-L154

And then have a separated class just for the "files":
<img width="858" alt="image" src="https://github.com/user-attachments/assets/b5ce17a7-545e-429d-b00d-f90d373dd5cb">

I renamed therefore some of your variables to make abap2UI5 look just as if its generated with the default fiori elements freestyle setup, e.g. renamed rootView to App and abap2UI5 is now in View1.

Is there any reason why you used a flexbox here:
https://github.com/mzinnoecker/abap2UI5/blob/da5578f589801d803aceed5e410b48cb2d126682/src/01/02/z2ui5_cl_core_http_get.clas.abap#L725

i switched it now to App and also needed to change the former add item:
https://github.com/mzinnoecker/abap2UI5/blob/da5578f589801d803aceed5e410b48cb2d126682/src/01/02/z2ui5_cl_core_http_get.clas.abap#L614

To insert page:
https://github.com/abap2UI5/abap2UI5/blob/test-components-views/src/01/02/z2ui5_cl_core_http_get.clas.locals_imp.abap#L563

If there was a reason for having a flexbox and addItem, just let me know.

**Comment by mzinnoecker** [comment]
Thats unfortunate :( anyone having good connections to SAP dev to ask for releasing the method for setting stateful? 😃 even SAP having not everything stateless in public cloud, would be just fair 😃 

 like it!
No specific reason for FlexBox instead of App. Just tried out things and that was the very first try. Changing to App is just as good. 👍
Also did not add initial routing as there was no need for… we can think of adding some routing things later 😃 
Thanks for the change and merging. Good teamwork!

**Comment by oblomov-dev** [comment]
yes its always a bit unfair in cloud, sap can just use everything :)

great! 💪

i merged everything to main now, made a lot of refactoring and also adjusted:
https://github.com/abap2UI5-connectors/fiori-launchpad-connector-onpremise

the idea is that both repos look the same and we can use the connector as a vscode abap2UI5 skeleton as you asked in the beginning. then we just need to copy&paste the code and bring the functionality 1:1 to abap2UI5 to the new "simulated" path and file structure. still some refactoring to do but we will get to it.

all in all great work and i think the changes are a big step forward! all futures features with navigation etc. can be integrated a lot easier now i hope!

**Comment by mzinnoecker** [comment]
Hey,
that‘s really nice! Like it very much 👍 
Will write a few words as a start tomorrow :)

**Comment by choper725** [comment]
@oblomov-dev this is the new handler needed ?

btw, this means variant management should work now if implemented?

**Comment by oblomov-dev** [comment]
@choper725 
i think you can still use the "old" handler. the new logic will also work with it. i have in mind to do some refactoring in the upcoming days and maybe integrate the new stateful logic into it, so some breaking changes are on the way and you can better wait for it before you switch.

yes basically everything is now setup as if it is a normal UI5 app. just use this class:
https://github.com/abap2UI5/abap2UI5/blob/main/src/01/02/z2ui5_cl_core_ui5_app.clas.abap
and add your JS/JSON there. variant management should work but never tried, so feel free to play around with it.

made some bigger changes to the get handler & index.html, unfortunately needed also to change the SizeLimit parameter. i  have to keep the JS now static so i deleted all dynamic injections into the index.html. check this sample:
https://github.com/abap2UI5/abap2UI5-samples/blob/main/src/z2ui5_cl_demo_app_071.clas.abap
if needed we can also try to implement it differently, its only important that it is not done on index.html level to simply maintenance etc.

**Comment by choper725** [comment]
ok, ill wait till changes made, to play around with it

thx fir the tips!

**Comment by oblomov-dev** [comment]
this issue got big, i will close this one, just create a new one if new issues come up.

updated now to a new handler. the whole stateful logic is now integrated into the handler and no extra coding needs to be copied for end-users anymore. but unfortunately breaking changes, but i think we are more future ready with this change. 

Feeback & Bugs here:
https://github.com/abap2UI5/abap2UI5/issues/1447

@mzinnoecker will probably do a linkedin post about statefulness, can i refer to you as a contributor or do you prefer to stay anonymous?

**Comment by mzinnoecker** [comment]
Like it!
Yes let's close this issue. Stateful should be done 👍 

And yes, just feel free! Thanks for asking
